components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/dao/impl/IdentityUserMetadataMgtDAOImpl.java

```diff
@@ -0,0 +1,508 @@
+import org.wso2.carbon.user.api.RealmConfiguration;
+import org.wso2.carbon.user.core.UserCoreConstants;
+import org.wso2.carbon.user.core.UserStoreException;
+import org.wso2.carbon.user.core.UserStoreManager;
+import org.wso2.carbon.user.core.claim.ClaimManager;
+import org.wso2.carbon.user.core.service.RealmService;
+import org.wso2.carbon.user.core.util.DatabaseUtil;
+import org.wso2.carbon.user.core.util.UserCoreUtil;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.Locale;
+import java.util.Map;
+import javax.sql.DataSource;
+
+/**
+ * Implementation class for IdentityUserMetadataMgtDAO.
+ */
+public class IdentityUserMetadataMgtDAOImpl implements IdentityUserMetadataMgtDAO {
+
+    private static final Log log = LogFactory.getLog(IdentityUserMetadataMgtDAOImpl.class);
+
+    private final IdentityDataStoreCache identityDataStoreCache = IdentityDataStoreCache.getInstance();
+
+    @Override
+    public void updateUserMetadata(UserStoreManager userStoreManager, String username,
+                                   String claimURI, String value, String eventName) throws IdentityEventException {
+
+        if (username.contains(IdentityRecoveryConstants.TENANT_ASSOCIATION_MANAGER)) {
+            return;
+        }
+
+        String tenantId;
+        try {
+            tenantId = String.valueOf(userStoreManager.getTenantId());
+        } catch (UserStoreException e) {
+            throw new IdentityEventException(
+                    String.format("Error occurred while retrieving the tenantId of the user related to %s event.",
+                            eventName), e);
+        }
+
+        String userStoreDomain = UserCoreUtil.extractDomainFromName(username);
+        if (StringUtils.isNotBlank(userStoreDomain) && IdentityRecoveryConstants.ASGARDEO_USER_DOMAIN_NAME
+                .equalsIgnoreCase(userStoreDomain) && isStoreIdentityClaimsInUserStoreEnabled(userStoreManager)) {
+            username = UserCoreUtil.removeDomainFromName(username);
+            storeMetadataToUserStore(userStoreManager, username, tenantId, claimURI, value, eventName);
+            return;
+        }
+
+        storeMetadataToIdentityStore(username, tenantId, claimURI, value, eventName);
+    }
+
+    @Override
+    public UserIdentityClaim loadUserMetadataFromCache(UserStoreManager userStoreManager, String userName) {
+
+        try {
+            if (userName != null) {
+                userName = UserCoreUtil.removeDomainFromName(userName);
+                if (!IdentityUtil.isUserStoreCaseSensitive(userStoreManager)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Case insensitive user store found. Changing username from : " + userName +
+                                " to : " + userName.toLowerCase(Locale.ENGLISH));
+                    }
+                    userName = userName.toLowerCase(Locale.ENGLISH);
```

Comment:
Shall we move this outside of the if block? So we can avoid calling it repeatedly.

Reply:
Fixed in https://github.com/wso2-extensions/identity-governance/pull/697/commits/dd3419fd69cc40e847e40ed9dfb29c867a3f0018